### PR TITLE
feature #115: Implemented an admin comments flow

### LINKED ISSUE #115: [FEATURE] Create an admin answers flow
We need to create "Search answers" and "Edit answer" pages

## Front-end part

Search answers:

- Sidebar that is already implemented
- Search panel:
  - Search input that searches by text of answer
  - Filter by semester and year in the same multichoice dropdown (given or taken)
  - Sorting dropdown with variants "by teacher", "by semester", "by subject"
  - Icon button to change the direction of sorting (asc/desc)
- Table:
  - Text (string)
  - Teacher (Last name and initials)
  - Subject (string)
  - Semester and year
  - Button "Edit" which opens "Edit answer" page
  - Button "Delete" which opens confirmation pop-up
  - Below the table there are settings for page size and page (from MUI component)
  
image

Edit answer:

- Sidebar that is already implemented
- Button "Cancel" which opens "Search grants" page
- Button "Delete" which opens confirmation pop-up. After deleting opens "Search grants" page
- Button "Save" which send a request on back-end endpoint which saves data. After saving opens "Search grants" page
- Input for text (min 4, max 4000)

image

## Back-end part

- Endpoint for getting answers (restricted access) with ability to:
  - filter by:
    - semester and year
  - and sort by:
    - teacher
    - subject
    - semester and year
  - that returns:
    - text
    - teacher info
    - subject info
    - semester and year 
    - user id
    - discipline teacher id
    - question id

- Endpoint for editing answer (by discipline teacher id, question id and user id):
  - text (string, min 4, max 4000)

### PR DESCRIPTION
**Pull request by JessFreak**
back-end part:
- created CommentByQuestionIdPipe
- created Comments responses
- created getAllComments endpoint
- edited update and delete comment endpoints
- edited getComment in QuestionMapper
- fixed some mistakes

front-end part:
- implemented page for an admin comments flow
- added async functions and types to get all comments and delete them
- added search panel
- added comments table
- implemented edit comments page

closes https://github.com/fictadvisor/fictadvisor/issues/115

### CODE REVIEW

**Review by illiakroshka — DISMISSED** [review]

File: fictadvisor-api/src/v2/api/controllers/DisciplineTeacherController.ts [hunk 1]
```diff
@@ -329,8 +386,9 @@ export class DisciplineTeacherController {
       QuestionId should not be empty`,
   })
   @ApiForbiddenResponse({
```
Comment (PavloVlasiuk) [inline fictadvisor-api/src/v2/api/controllers/DisciplineTeacherController.ts:388]: You have used ApiForbiddenResponse decorator twice, remove one

File: fictadvisor-api/src/v2/api/dtos/QueryAllCommentsDTO.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  subject = 'subject',
+  semester = 'semester',
+}
+
+export class QueryAllCommentsDTO extends QueryAllDTO {
+  @ApiPropertyOptional({
+    enum: SortCommentsEnum,
+    default: SortCommentsEnum.subject,
+  })
+  @IsEnum(SortCommentsEnum, validationOptionsMsg('Sort should be an enum'))
```
Comment (PavloVlasiuk) [inline fictadvisor-api/src/v2/api/dtos/QueryAllCommentsDTO.ts:19]: we use pattern to name exceptions in such way: "Some value" must be a "type". So rename this exception according to the pattern

File: fictadvisor-api/src/v2/api/dtos/QueryAllCommentsDTO.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    sort?: string;
+
+  @ApiPropertyOptional({
+    description: 'Array of studying semesters',
+  })
+  @IsOptional()
+  @ValidateNested(validationOptionsMsg('Each value of semesters should be an studying semester', true))
```
Comment (PavloVlasiuk) [inline fictadvisor-api/src/v2/api/dtos/QueryAllCommentsDTO.ts:26]: change this exception too

File: fictadvisor-api/src/v2/api/dtos/QueryAllDisciplinesDTO.ts [hunk 4]
```diff
@@ -16,9 +16,7 @@ export class QueryAllDisciplinesDTO extends QueryAllDTO {
     description: 'Array of studying semesters',
   })
   @IsOptional()
-  @ValidateNested({
-    each: true,
-  })
+  @ValidateNested(validationOptionsMsg('Each value of semesters should be an studying semester', true))
```
Comment (PavloVlasiuk) [inline fictadvisor-api/src/v2/api/dtos/QueryAllDisciplinesDTO.ts:19]: and this

File: fictadvisor-api/src/v2/api/dtos/QueryAllStudentDTO.ts [hunk 5]
```diff
@@ -38,7 +38,7 @@ export class QueryAllStudentDTO extends QueryAllDTO {
     description: 'Array of groups ids',
   })
   @IsArray(validationOptionsMsg('Groups should be an array'))
```
Comment (PavloVlasiuk) [inline fictadvisor-api/src/v2/api/dtos/QueryAllStudentDTO.ts:40]: and this

File: fictadvisor-api/src/v2/api/services/DisciplineTeacherService.ts [hunk 6]
```diff
@@ -231,26 +232,9 @@ export class DisciplineTeacherService {
...
+    const discipline = await this.disciplineRepository.findById(disciplineId);
+    const dbRoles = await this.getDbRoles(discipline, roles);
 
-      dbRoles.push({
-        role,
-        disciplineTypeId: disciplineType.id,
-      });
-    }
     return this.disciplineTeacherRepository.create({
       teacherId: teacherId,
```
Comment (PavloVlasiuk) [inline fictadvisor-api/src/v2/api/services/DisciplineTeacherService.ts:239]: Remove unnecessary key: value pairs as they have same names

**Review by PavloVlasiuk — CHANGES_REQUESTED** [review]

File: fictadvisor-web/src/components/pages/admin/admin-comments/pages/edit-comments-page/CommentsAdminEditPage.tsx [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                text="Зберегти"
+                sx={styles.button}
+                type="submit"
+              />
+            </Stack>
+          </Box>
+          <Box sx={styles.body}>
+            <Box sx={styles.textArea}>
+              <TextArea
+                name={'answer'}
```
Comment (Markusplay) [inline fictadvisor-web/src/components/pages/admin/admin-comments/pages/edit-comments-page/CommentsAdminEditPage.tsx:107]: remove curly braces

File: fictadvisor-web/src/components/pages/admin/admin-comments/pages/edit-comments-page/CommentsAdminEditPage.styles.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+export const body: SxProps<Theme> = {
+  display: 'flex',
+  gap: '36px',
+  padding: '24px 16px',
+  '.css-1qp9ii-MuiInputBase-root-MuiOutlinedInput-root, .css-1od1924-MuiFormLabel-root-MuiInputLabel-root':
+    {
+      marginTop: 0,
+      height: 'max-content',
+    },
+};
```
Comment (Markusplay) [inline fictadvisor-web/src/components/pages/admin/admin-comments/pages/edit-comments-page/CommentsAdminEditPage.styles.ts:39]: do not use hashed styles, use MUI Api to set needed styles

File: fictadvisor-web/src/components/pages/admin/admin-comments/pages/edit-comments-page/CommentsAdminEditPage.styles.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+export const title: SxProps<Theme> = {
+  borderBottom: '1px solid',
+  borderColor: 'backgroundDark.400',
+  padding: '16px',
+  width: '60%',
+  '& .MuiCardHeader-subheader': {
+    typography: 'body1',
+    color: 'grey.500',
+    display: '-webkit-box',
```
Comment (Markusplay) [inline fictadvisor-web/src/components/pages/admin/admin-comments/pages/edit-comments-page/CommentsAdminEditPage.styles.ts:18]: we don't need to use vendor prefix

File: fictadvisor-web/src/components/common/ui/form/text-area-mui/TextArea.tsx [hunk 10]
```diff
@@ -23,7 +23,9 @@ interface TextAreaProps {
   disabled?: boolean;
   showRemark?: boolean;
   rowsNumber?: number;
+  value?: string;
   sx?: SxProps<Theme>;
+  onChange?: (e: any) => void;
```
Comment (Markusplay) [inline fictadvisor-web/src/components/common/ui/form/text-area-mui/TextArea.tsx:28]: don't use `any` for type

File: fictadvisor-web/src/components/pages/admin/admin-comments/components/admin-comments-search/AdminCommentsSearch.styles.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+export const button: SxProps<Theme> = {
+  width: '135px',
+  height: '48px',
+  borderRadius: '8px',
+};
+
+export const divider: SxProps<Theme> = {
+  height: '46px',
+  width: '1px',
+  background: '#747474',
```
Comment (Markusplay) [inline fictadvisor-web/src/components/pages/admin/admin-comments/components/admin-comments-search/AdminCommentsSearch.styles.ts:38]: use color from theme

File: fictadvisor-web/src/components/pages/admin/admin-comments/components/admin-comments-search/constants/CommentsInitialValues.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import { CommentsAdminSearchFormFields } from '../types/AdminCommentsSearch';
+
+export const InitialValues: CommentsAdminSearchFormFields = {
```
Comment (Markusplay) [inline fictadvisor-web/src/components/pages/admin/admin-comments/components/admin-comments-search/constants/CommentsInitialValues.ts:3]: should be camel case

File: fictadvisor-web/src/components/pages/admin/admin-comments/components/admin-comments-table/CommentsTable.styles.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  borderBottom: '1px solid',
+  borderColor: 'backgroundDark.400',
+  width: 'fit-content',
+};
+
+export const tableColumn: SxProps<Theme> = {
+  flexDirection: 'row',
+  gap: '8px',
+  alignItems: 'center',
+  display: '-webkit-box',
```
Comment (Markusplay) [inline fictadvisor-web/src/components/pages/admin/admin-comments/components/admin-comments-table/CommentsTable.styles.ts:22]: we don't need vendor prefix

**Review by Markusplay — CHANGES_REQUESTED** [review]

File: fictadvisor-web/src/components/pages/admin/admin-comments/pages/edit-comments-page/CommentsAdminEditPage.styles.ts (hunk 9 above)
Comment (nnstsya) [reply fictadvisor-web/src/components/pages/admin/admin-comments/pages/edit-comments-page/CommentsAdminEditPage.styles.ts:18]: it does not display answers correctly without it

**Review by igaryakqwe — APPROVED** [review]

**Review by PavloVlasiuk — APPROVED** [review]

File: fictadvisor-api/src/v2/api/dtos/QueryAllCommentsDTO.ts [hunk 14]
```diff
@@ -0,0 +1,29 @@
+import { validationOptionsMsg } from '../../utils/GLOBALS';
+import { ApiPropertyOptional } from '@nestjs/swagger';
+import { IsEnum, IsOptional, ValidateNested } from 'class-validator';
+import { Type } from 'class-transformer';
+import { QueryAllDTO } from '../../utils/QueryAllDTO';
+import { QuerySemesterDTO } from './QuerySemesterDTO';
+
+export enum SortCommentsEnum {
+  teacher = 'teacher',
```
Comment (f-ogoog) [inline fictadvisor-api/src/v2/api/dtos/QueryAllCommentsDTO.ts:9]: write keys with upper case

File: fictadvisor-api/src/v2/api/responses/PaginatedCommentsResponse.ts [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+
+export class PaginatedCommentsResponse {
+  @ApiProperty({
+    description: 'question answers with TEXT type (comments)',
+    type: [CommentResponse],
+  })
+    comments: CommentResponse[];
+
+  @ApiProperty({
+    description: 'Pagination parameters',
```
Comment (f-ogoog) [inline fictadvisor-api/src/v2/api/responses/PaginatedCommentsResponse.ts:14]: add type

**Review by f-ogoog — DISMISSED** [review]

File: fictadvisor-web/src/app/admin/comments/edit/[comment]/page.tsx [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+import CommentsAdminEditPage from '@/components/pages/admin/admin-comments/pages/edit-comments-page';
+
+interface AdminCommentEditProps {
+  params: {
+    comment: string;
+  };
+}
+
+const AdminCommentEdit: FC<AdminCommentEditProps> = ({ params }) => {
+  const comment = JSON.parse(decodeURIComponent(params.comment));
```
Comment (igaryakqwe) [inline fictadvisor-web/src/app/admin/comments/edit/[comment]/page.tsx:13]: you can use params.comment without decoding and parsing

File: fictadvisor-web/src/components/pages/admin/admin-comments/components/admin-comments-table/CommentsTable.tsx [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    } catch (e) {
+      if (isAxiosError(e)) {
+        toastError.displayError(e.response?.data.message);
+      }
+    }
+  };
+
+  return (
+    <Table>
+      <thead>
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-comments/components/admin-comments-table/CommentsTable.tsx:63]: MUI provides the TableHead component, use it instead common thead

**Review by igaryakqwe — DISMISSED** [review]

**Review by f-ogoog — APPROVED** [review]

**Pull request merged**